packages/core/src/middleware-http-auth-scheme/resolveAuthOptions.ts

```diff
@@ -0,0 +1,39 @@
+import { HttpAuthOption } from "@smithy/types";
+
+/**
+ * Resolves list of auth options based on the supported ones, vs the preference list.
+ *
+ * @param candidateAuthOptions list of supported auth options selected by the standard
+ *   resolution process (model-based, endpoints 2.0, etc.)
+ * @param authSchemePreference list of auth schemes preferred by user.
+ * @returns
+ */
+export const resolveAuthOptions = (
+  candidateAuthOptions: HttpAuthOption[],
+  authSchemePreference: string[]
+): HttpAuthOption[] => {
+  if (!authSchemePreference || authSchemePreference.length === 0) {
+    return candidateAuthOptions;
+  }
+
+  // reprioritize candidates based on user's preference
+  const preferredAuthOptions = [];
+
+  for (const preferredSchemeName of authSchemePreference) {
+    for (const candidateAuthOption of candidateAuthOptions) {
+      const candidateAuthSchemeName = candidateAuthOption.schemeId.split("#")[1];
+      if (candidateAuthSchemeName === preferredSchemeName) {
+        preferredAuthOptions.push(candidateAuthOption);
+      }
+    }
+  }
+
+  // add any remaining candidates that weren't in the preference list
+  for (const candidateAuthOption of candidateAuthOptions) {
+    if (!preferredAuthOptions.find(({ schemeId }) => schemeId === candidateAuthOption.schemeId)) {
+      preferredAuthOptions.push(candidateAuthOption);
+    }
+  }
+
+  return preferredAuthOptions;
```

Comment:
because the lists are so small, I feel that this is optional, but you can do this without nested for loops like this:

```ts
// runnable in console
{
const candidates = [{ id: 'a', }, { id: 'b' }];
const preference = ['c', 'b'];

const candidateMap = candidates.reduce((map, item) => {
  map.set(item.id, item);
  return map;
}, new Map);

const preferredMap = new Map;

for (const pref of preference) {
  if (candidateMap.has(pref)) {
    preferredMap.set(pref, candidateMap.get(pref));
  }
}

for (const [id, item] of candidateMap) {
  preferredMap.set(id, item);
}

console.log([...preferredMap.values()])
}
```

Reply:
I think we should retain the resolve function, as it's recommended in the internal cross SDK proposal.
Also, the JS SDK implementation uses map for authSchemes which do not have order preserved. The order is preserved in authOptions instead.

Reply:
I'm not sure if you're referring to the Map in the example, but those retain insertion order like modern objects. 

But let's keep the function as is since it is in the design.